File: common-client-1_21_8/src/main/java/com/inotsleep/dreamdisplays/client_1_21_8/packets/PremiumPacket.java
package com.inotsleep.dreamdisplays.client_1_21_8.packets;

import com.inotsleep.dreamdisplays.client_1_21_8.DreamDisplaysClientCommon;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import org.jetbrains.annotations.NotNull;


/**
 * Sent from the server to the client to indicate whether the user has a premium.
 * This packet is used to enable or disable premium features (in the client)
 * @param premium true if the user has a premium, false otherwise.
 */
public record PremiumPacket(boolean premium) implements CustomPacketPayload {
    public static final Type<PremiumPacket> PACKET_ID =
            new Type<>(ResourceLocation.fromNamespaceAndPath(DreamDisplaysClientCommon.MOD_ID, "premium"));

    public static final StreamCodec<FriendlyByteBuf, PremiumPacket> PACKET_CODEC =
            StreamCodec.of(
                    (buf, packet) -> buf.writeBoolean(packet.premium()),
                    (buf) -> {
                        boolean premium = buf.readBoolean();
                        return new PremiumPacket(premium);
                    });

    @Override
    public @NotNull Type<? extends CustomPacketPayload> type() {
        return PACKET_ID;
    }
}

File: common/src/main/java/com/dreamdisplays/net/PremiumPacket.kt
package com.dreamdisplays.net

import com.dreamdisplays.Initializer
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.Identifier
import org.jspecify.annotations.NullMarked

/**
 * Packet for premium status (1080p+ quality, etc.)
 */
@NullMarked
@JvmRecord
data class PremiumPacket(val premium: Boolean) : CustomPacketPayload {
    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> {
        return PACKET_ID
    }

    companion object {
        @JvmField
        val PACKET_ID: CustomPacketPayload.Type<PremiumPacket> = CustomPacketPayload.Type<PremiumPacket>(
            Identifier.fromNamespaceAndPath(
                Initializer.MOD_ID, "premium"
            )
        )

        @JvmField
        val PACKET_CODEC: StreamCodec<FriendlyByteBuf, PremiumPacket> = StreamCodec.of<FriendlyByteBuf, PremiumPacket>(
            { buf: FriendlyByteBuf?, packet: PremiumPacket? ->
                ByteBufCodecs.BOOL.encode(
                    buf!!,
                    packet!!.premium
                )
            },
            { buf: FriendlyByteBuf? ->
                val premium = ByteBufCodecs.BOOL.decode(buf!!)
                PremiumPacket(premium)
            })
    }
}
